From c188b176d35f63c3a0dffb89c0bebbd8ce0457a0 Mon Sep 17 00:00:00 2001
From: Jordan Crouse <jcrouse@codeaurora.org>
Date: Wed, 28 Dec 2011 00:37:23 +0100
Subject: [PATCH] msm: kgsl: Fixup external memory offsets

Offsets passed in for external memory chunks are not always
4K aligned, so make sure the differences are accounted for.

CRs-fixed: 290267
Change-Id: I3c11b7f853679e00af8ebc9c7f54443d2fe6ec6d
Signed-off-by: Jordan Crouse <jcrouse@codeaurora.org>
---
 drivers/video/msm/gpu/kgsl_adreno205_hc/kgsl.c |   11 ++++++++++-
 1 files changed, 10 insertions(+), 1 deletions(-)

diff --git a/drivers/video/msm/gpu/kgsl_adreno205_hc/kgsl.c b/drivers/video/msm/gpu/kgsl_adreno205_hc/kgsl.c
index 2e80340..1731915 100644
--- a/drivers/video/msm/gpu/kgsl_adreno205_hc/kgsl.c
+++ b/drivers/video/msm/gpu/kgsl_adreno205_hc/kgsl.c
@@ -749,6 +749,8 @@ static int kgsl_open(struct inode *inodep, struct file *filep)
 
 	BUG_ON(private == NULL);
 
+	gpuaddr &= PAGE_MASK;
+
 	list_for_each_entry(entry, &private->mem_list, list) {
 		if (entry->memdesc.gpuaddr == gpuaddr) {
 			result = entry;
@@ -1296,6 +1298,9 @@ static int kgsl_setup_phys_file(struct kgsl_mem_entry *entry,
 	if (size == 0)
 		size = len;
 
+	/* Adjust the size of the region to account for the offset */
+	size += offset & ~PAGE_MASK;
+
 	size = ALIGN(size, PAGE_SIZE);
 
 	if (_check_region(offset & PAGE_MASK, size, len)) {
@@ -1357,6 +1362,9 @@ static int kgsl_setup_hostptr(struct kgsl_mem_entry *entry,
 	if (size == 0)
 		size = len;
 
+	/* Adjust the size of the region to account for the offset */
+	size += offset & ~PAGE_MASK;
+
 	size = ALIGN(size, PAGE_SIZE);
 
 	if (_check_region(offset & PAGE_MASK, size, len)) {
@@ -1505,7 +1513,8 @@ static long kgsl_ioctl_map_user_mem(struct kgsl_device_private *dev_priv,
 	if (result)
 		goto error_put_file_ptr;
 
-	param->gpuaddr = entry->memdesc.gpuaddr;
+	/* Adjust the returned value for a non 4k aligned offset */
+	param->gpuaddr = entry->memdesc.gpuaddr + (param->offset & ~PAGE_MASK);
 
 	entry->memtype = KGSL_EXTERNAL_MEMORY;
 
-- 
1.7.5.4

